docs/rules/lines-between-class-members.md

```diff
@@ -0,0 +1,152 @@
+# enforce lines between class members (lines-between-class-members)
+
+This rule improves readability by enforcing lines between class members.
+
+## Rule Details
+
+Examples of **incorrect** code for this rule:
+
+```js
+/* eslint lines-between-class-members: ["error", "always"]*/
+class MyClass {
+  foo() {
+    //...
+  }
+  bar() {
+    //...
+  }
+}
+```
+
+Examples of **correct** code for this rule:
+
+```js
+/* eslint lines-between-class-members: ["error", "always"]*/
+class MyClass {
+  foo() {
+    //...
+  }
+
+  bar() {
+    //...
+  }
+}
+```
+
+### Options
+
+This rule has one option, which can be a string option or an object option.
+
+String option:
+
+* `"always"`(default) requires line breaks between methods
+* `"never"` disallows line breaks between methods
+
+Object option:
+
+* `"multiline": "always"` requires line breaks if the method is multiline
+* `"multiline": "never"` disallow line breaks if the method is multiline
+* `"singleline": "always"` requires line breaks if the method is singleline
+* `"singleline": "never"` disallow line breaks if the method is singleline
```

Comment:
I'm a bit confused about how these options are supposed to work. The rule checks whether there are linebreaks between two methods, so which method does the rule look at when checking whether the method is multiline? For example:

```js
/* eslint lines-between-class-members: [error, {multiline: always, singleline: never}] */

class Foo {
    bar() {
    }
    baz() {}
}
```

Reply:
hmm..for now it only check the first method to decide if it is multi.
I see, it might be better to check both. eg:
```js
/* eslint lines-between-class-members: [error, {multiline: always}] */
// valid
class Foo {
    bar() {
    }

    baz() {
    }
}

/* eslint lines-between-class-members: [error, {singleline: always}] */
// valid
class Foo {
    bar() {}

    baz() {}
}
```

Reply:
if need, we can add a new option to enforce between multi and single.(I think this could be done in another PR)

Reply:
My preference would be to get this figured out in this PR to minimize the chance of breaking changes in the future.

Reading over the original issue again, I see the confusion. The original JSCS rules only enforce lines _after_ class members, which is why this configuration schema was suggested. My vote would be to keep the behavior implemented in this PR and rename the rule to padding-line-after-class-members.

Reply:
Would the rule also enforce padding lines after the last class member?

```js
class Foo {
    bar() {}
                    // <-- line here?
}
```

Reply:
No, I believe that is padded-block's responsibility (since it could cause a rule conflict). In other cases like this we have always had the rue ignore white space after the opening brace and before the closing brace.

Reply:
Makes sense to me. In that case, I think `padding-lines-between-*` would be a better name than `padding-lines-after-*`, since the latter would imply that the rule would check that case.

Reply:
@not-an-aardvark I'm not sure I'm following. I think we should leave `padding-line-between-statements` as is because it does actually evaluate the statement before and after. This rule (at least in its current implementation, following the accepted proposal) only checks if a newline should occur after a class member. Are you advocating for `padding-line-between-class-members`?

Reply:
Yes, I'm advocating for `padding-line-between-class-members`. I think the following code contains a padding line after a class member:

```js
class Foo {
    bar() {}
                    // <-- line
}
```

If the new rule isn't checking that case, it seems misleading to call it `padding-line-after-class-members`.

Reply:
I see your point. This is a convention we've followed in a number of rules (though I think many of them were deprecated by the newer padding-line-between-statements rule). In my mind, having the name of the rule call out the majority use case provides more clarity, since it describes more accurately what the rule is actually checking.

I'm still advocating for padding-line-after, but if we do go the padding-line-between route, I think the current configuration options are confusing. Could it maybe be better to have the option names change to afterSingleLine and afterMultiLine?

Reply:
👍 to @kaicataldo `padding-line-after-class-members` is more clear, and just only one exception.

Reply:
> In my mind, having the name of the rule call out the majority use case provides more clarity, since it describes more accurately what the rule is actually checking.

I'm confused by this point; it seems like `between` describes what the rule checks for in all cases, and `after` describes what the rule checks for in all cases except one. Doesn't this make `between` a more accurate description of what the rule is checking?

Reply:
hmm..seems also reasonable. 😂 

But I'm wondering the options -- `multiline` and `singleline`, are a bit confusing for  `padding-line-between-class-members`; for `padding-line-after`, it is more straightforward.

Reply:
@Aladdin-ADD explained what I was trying to get at better than I did! The configuration options are what are confusing to me (for the same reason you were asking about in [this comment](https://github.com/eslint/eslint/pull/9141#discussion_r137139744).

Reply:
What if we called the options something like `afterSingleLine` and `afterMultiLine`?

(As a sidenote, the value of these options isn't entirely clear to me. I realize we need to separate `singleLine` behavior for JSCS compatibility, but I think maybe we should just add the minimal API necessary to do that, rather than changing the rule name to accommodate the option names.)

Reply:
I do think that's a better option than the current configuration names. To be clear though, we do make these exceptions for compatibility with padded-blocks in at least one other [current rule](https://github.com/eslint/eslint/blob/master/lib/rules/lines-around-directive.js#L139).

> As a sidenote, the value of these options isn't entirely clear to me. I think maybe we should just add the minimal API necessary to do that, rather than changing the rule name to accommodate the option names.

My motivation in suggesting a new rule name wasn't only to work around the option names - I do think it makes it clearer as to what the rule is actually checking (with the caveat that I have the tribal knowledge of knowing that we make exceptions for `padded-blocks` and in the past have deferred to that rule for enforcing those styles).

That being said, I agree that the API feels a bit YAGNI. What if we added one explicit option called something like `exceptSingleLineMembers` that would disallow newlines after single line members when the rule is configured with `"always"` (and would be ignored when set to `"never"`). This would allow us to better match the JSCS API and I think is easier to understand from a configuration perspective.

```
{
    "lines-after-class-members": ["error", "always" or "never", { 
        exceptSingleLineMembers: true or false
    }]
}
```

@Aladdin-ADD Thanks for your patience and input as we figure this out!

Reply:
I see your point. I suppose the issue is that even if we say that we're making an exception to the rule to avoid conflicts with `padded-blocks`, we're still introducing a case where the rule would not be doing what it says on the tin. That seems like a potential source of confusion, and I think we could avoid it by making the rule's name match its actual behavior.

I would rather have a slightly less consistent schema that is clear and correct in all cases, rather than having a slightly simpler schema that incorrectly describes a common case.

Reply:
@not-an-aardvark Makes sense. Mind summarizing what you're advocating for now? I'd love to hash this out ASAP so we don't hold this up any longer!

Reply:
I think we should call it `padding-lines-between-class-members`, and call the options something like `afterSingleLine` and `afterMultiLine` (or if we're using only one option, call it `exceptAfterSingleLine`).

Reply:
That works for me! I do think having a single option is clearer and less cumbersome. What do you think @not-an-aardvark  @Aladdin-ADD?

Reply:
agree!

Reply:
:+1:

Reply:
Great. Does that all make sense @Aladdin-ADD? It sounds like we want the configuration to be:
```
{
    "lines-between-class-members": ["error", "always" or "never" (defaulting to "always"), { 
        exceptAfterSingleLine: true or false (defaulting to true)
    }]
}
```

I think we should only use the `exceptAfterSingleLine` option when the rule is configured with "always".

Actually, sorry, one final thing to work out - does this option ignore single line methods (i.e. it can have either a single line or not, which I believe is what the JSCS rule does), or does it enforce no newline after them?

I ask because I can imagine situations where someone might want to group single line members together but have a newline between the group and the methods following it. Or, in the future, for types vs class properties such as in [this Vue Class Component example](https://github.com/vuejs/vue-class-component#example).

Reply:
Whoops, wrong example. Sorry! Here's the [correct one](https://github.com/vuejs/vue-class-component/blob/master/example/App.vue#L22).

Reply:
one more question:  you are saying if `exceptAfterSingleLine`  is  set to `false`, just ignore the member? 
what if a user want a opposite config (not just ignore)?

Reply:
I guess I got your point. maybe it is more configurable to adopt plan A in  https://github.com/eslint/eslint/pull/9141#discussion_r137960210

Reply:
> you are saying if exceptAfterSingleLine is set to false, just ignore the member

If we just want to support JSCS compatibility, I think that if `exceptAfterSingleLine` is set to `false` it should not make an exception for single line members and should treat them the same as the multi line members. 

Oops, I see the confusion (I accidentally had listed `true` as the default for `exceptAfterSingleLine`). It should definitely be `false` by default, so you would only enable it when the rule is configured with `"always"` and you want to create an exception for single line members.

Reply:
>Actually, sorry, one final thing to work out - does this option ignore single line methods (i.e. it can have either a single line or not, which I believe is what the JSCS rule does), or does it enforce no newline after them?

```js
/* eslint lines-between-class-members": ["error", "always", {"exceptAfterSingleLine": true}]*/
class Foo{
  bar(){}
                    // <-- line here?
  baz(){}
}
```
surely the last question, so what is your idea on this?

Reply:
>Actually, sorry, one final thing to work out - does this option ignore single line methods (i.e. it can have either a single line or not, which I believe is what the JSCS rule does), or does it enforce no newline after them?

```js
/* eslint lines-between-class-members": ["error", "always", {"exceptAfterSingleLine": true}]*/
class Foo{
  bar(){}
                    // <-- line here?
  baz(){}
}
```
surely the last question, so what is your idea on this?

Reply:
I'm wondering whether single line members are ignored or if we actually enforce no line breaks.

If it's ignored, this would not warn:
```js
/* eslint "lines-between-class-members": ["error", "always", {"exceptAfterSingleLine": true}]*/
class Foo{
  bar(){}

  baz(){}
}
```